populator-machinery/controller.go

```diff
@@ -527,13 +556,49 @@ func (c *controller) syncPvc(ctx context.Context, key, pvcNamespace, pvcName str
 	}
 
 	// Look for the populator pod
-	podName := fmt.Sprintf("%s-%s", populatorPodPrefix, pvc.UID)
-	c.addNotification(key, "pod", c.populatorNamespace, podName)
 	var pod *corev1.Pod
-	pod, err = c.podLister.Pods(c.populatorNamespace).Get(podName)
-	if err != nil {
-		if !errors.IsNotFound(err) {
-			return err
+	podName := fmt.Sprintf("%s-%s", populatorPodPrefix, pvc.UID)
+	if c.usePod {
+		c.addNotification(key, "pod", c.populatorNamespace, podName)
+		pod, err = c.podLister.Pods(c.populatorNamespace).Get(podName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+
+	// If populate data without populator pod, and orginal StorageClass's VolumeBindingMode is VolumeBindingWaitForFirstConsumer,
+	// create a StorageClass with VolumeBindingImmediate for pvcPrime
+	storageClassPrimeName := *pvc.Spec.StorageClassName
+	if !c.usePod && storageClass.VolumeBindingMode != nil && storagev1.VolumeBindingWaitForFirstConsumer == *storageClass.VolumeBindingMode {
+		storageClassPrimeName = populatorStorageClassPrefix + "-" + *pvc.Spec.StorageClassName
+		scPrime, err := c.scLister.Get(storageClassPrimeName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+			c.addNotification(key, "sc", "", storageClassPrimeName)
+			bm := storagev1.VolumeBindingImmediate
+			scPrime = &storagev1.StorageClass{
+				ObjectMeta:           metav1.ObjectMeta{Name: storageClassPrimeName},
+				Provisioner:          storageClass.Provisioner,
+				Parameters:           storageClass.Parameters,
+				ReclaimPolicy:        storageClass.ReclaimPolicy,
+				MountOptions:         storageClass.MountOptions,
+				AllowVolumeExpansion: storageClass.AllowVolumeExpansion,
+				VolumeBindingMode:    &bm,
+				AllowedTopologies:    storageClass.AllowedTopologies,
+			}
+			if storageClass.Parameters != nil && storageClass.Parameters["volumeBindingMode"] != "" {
+				scPrime.Parameters["volumeBindingMode"] = string(bm)
+			}
+			scPrime, err = c.kubeClient.StorageV1().StorageClasses().Create(ctx, scPrime, metav1.CreateOptions{})
+			if err != nil {
+				return err
```

Comment:
If we are re-using the SC prime for multiple volumes of the same SC, then this should ideally return error of some already exists .

Best to register an informer and do a get first.

Reply:
Yes, I have the scPrime get first, if it already exists then skip, and if the err is NotFound create scPrime

Reply:
Thanks for pointing . I missed that